CMSgov/easi-app
# [EASI-3183] Expire LCID action

# EASI-3183

## Changes and Description

- Add a GraphQL mutation for manually expiring an LCID, `createSystemIntakeActionExpireLCID`.
- Add an `lcidStatus` field to the `SystemIntake` type in the GraphQL schema, with a resolver to calculate this status based on the existing fields on the system intake model. 
  - This field has possible values of `ISSUED`, `EXPIRED`, and `RETIRED`; only `ISSUED` and `EXPIRED` can currently be returned, `RETIRED` will be used in EASI-3179. It's added as part of this ticket to enable related frontend work; see [this Slack thread](https://cmsgov.slack.com/archives/CNU2B59UH/p1694551142068539).

## How to test this change

- Import the Postman collection from this branch.
- Create and submit an intake with the "SystemIntake Create" and "SUBMIT INTAKE (Form 4)" request under System Intake -> Initial Request.
- Try to expire the newly created intake's LCID with the "Expire LCID" request, under System Intake -> Admin Actions (v2) -> Expire LCID. This should return an error that the action is invalid because the intake doesn't have an LCID yet.
- Issue an LCID for the intake with the "Issue LCID, generating new LCID" request, under System Intake -> Admin Actions (v2) -> Issue LCID.
- Run the "SystemIntake Get (Decision)" query and check the value of the `lcidStatus` field - it should be `ISSUED`.
- Run the Expire LCID request again. This should be successful.
- Running the "SystemIntake Get (Decision)" query should now show that `lcidStatus` is set to `EXPIRED`, that `lcidExpiresAt` is set to midnight UTC on the current day, and `decisionNextSteps` should have the value "LCID expired; create a new LCID" from the Expire LCID request's input.
- Running the "SystemIntake Get (Feedback with LCID Change History)" query should now show an action and admin note from the Expire LCID request. The action's `lcidExpirationChange` field should show changes between `previousDate` and `newDate`, as well as `previousNextSteps` and `newNextSteps`.
- Run the Expire LCID request again. It should return an error that the action is invalid because the intake's LCID is already expired.

## Notes
- I moved the `getBaseLCIDAction()` utility function and its tests into a separate file, since they're now used by multiple functions in the `pkg/graph/resolvers/itgovactions/lcidactions/` folder.
- I started using the "Note" system for organizing repeated comments in `pkg/graph/resolvers/system_intake_actions.go`, adding `Note [Database calls from resolvers aren't atomic]`; see [this Slack discussion](https://oddball.slack.com/archives/C02G5D05PD4/p1694116864588119) for background. Let me know what you think of the concept and the syntax I'm using.
- I tried using the GraphQL `@deprecated` directive on our GQL schema for v1-only mutations, but that didn't seem to show any warnings or metadata in the GQL playground, Postman, or apollo codegen. Instead, I changed the comments on those mutations to use the `"""` documentation comment format, which makes those comments appear in the GQL playground, at least. (I didn't check Postman, apollo codegen, or writing queries with them in `src/queries/`)
